tiledb/sm/query/query_ast.h

```diff
@@ -0,0 +1,231 @@
+
+/** Represents a simple terminal/predicate **/
+class ASTNode {
+ public:
+  virtual ASTNodeTag get_tag() const {
+    return tag_;
+  }
+  virtual std::string to_str() const {
+    return "";
+  }
+  virtual ~ASTNode() {
+  }
+
+ private:
+  static const ASTNodeTag tag_{ASTNodeTag::NIL};
+};
+class ASTNodeVal : public ASTNode {
+  /** Value constructor. */
+ public:
+  ASTNodeVal(
+      std::string field_name,
+      const void* const condition_value,
+      const uint64_t condition_value_size,
+      const QueryConditionOp op)
+      : field_name_(field_name)
+      , op_(op) {
+    condition_value_data_.resize(condition_value_size);
+    condition_value_ = nullptr;
+
+    if (condition_value != nullptr) {
+      // Using an empty string litteral for empty string as vector::resize(0)
+      // doesn't guarantee an allocation.
+      condition_value_ =
+          condition_value_size == 0 ? (void*)"" : condition_value_data_.data();
+      memcpy(
+          condition_value_data_.data(), condition_value, condition_value_size);
+    }
+  }
+
+  /** Copy constructor. */
+  ASTNodeVal(const ASTNodeVal& rhs)
+      : field_name_(rhs.field_name_)
+      , condition_value_data_(rhs.condition_value_data_)
+      , condition_value_(
+            rhs.condition_value_ == nullptr ? nullptr :
+                                              condition_value_data_.data())
+      , op_(rhs.op_) {
+  }
+
+  /** Move constructor. */
+  ASTNodeVal(ASTNodeVal&& rhs)
+      : field_name_(std::move(rhs.field_name_))
+      , condition_value_data_(std::move(rhs.condition_value_data_))
+      , condition_value_(
+            rhs.condition_value_ == nullptr ? nullptr :
+                                              condition_value_data_.data())
+      , op_(rhs.op_) {
+  }
+
+  /** Assignment operator. */
+  ASTNodeVal& operator=(const ASTNodeVal& rhs) {
+    field_name_ = rhs.field_name_;
+    condition_value_data_ = rhs.condition_value_data_;
+    condition_value_ = rhs.condition_value_ == nullptr ?
+                           nullptr :
+                           condition_value_data_.data();
+    op_ = rhs.op_;
+
+    return *this;
+  }
+
+  /** Move-assignment operator. */
+  ASTNodeVal& operator=(ASTNodeVal&& rhs) {
+    field_name_ = std::move(rhs.field_name_);
+    condition_value_data_ = std::move(rhs.condition_value_data_);
+    condition_value_ = rhs.condition_value_ == nullptr ?
+                           nullptr :
+                           condition_value_data_.data();
+    op_ = rhs.op_;
+
+    return *this;
+  }
+
+  ASTNodeTag get_tag() const override {
+    return tag_;
+  }
+
+  std::string to_str() const override {
+    std::string result_str;
+    result_str = field_name_ + " " + query_condition_op_str(op_) + " ";
+    result_str += condition_value_data_.to_str();
+    return result_str;
+  }
+
+  /** The attribute name. */
+  std::string field_name_;
+
+  /** The value data. */
+  ByteVecValue condition_value_data_;
```

Comment:
Note story 15762 -- we will hopefully consolidate / simplify `condition_value_data_` and `condition_value_`